tfexec/cmd.go

```diff
@@ -275,3 +303,97 @@ func writeOutput(ctx context.Context, r io.ReadCloser, w io.Writer) error {
 		}
 	}
 }
+
+func (tf *Terraform) runTerraformCmd(ctx context.Context, cmd *exec.Cmd) error {
+	var errBuf strings.Builder
+
+	cmd.SysProcAttr = defaultSysProcAttr
+
+	// check for early cancellation
+	select {
+	case <-ctx.Done():
+		return ctx.Err()
+	default:
+	}
+
+	// Read stdout / stderr logs from pipe instead of setting cmd.Stdout and
+	// cmd.Stderr because it can cause hanging when killing the command
+	// https://github.com/golang/go/issues/23019
+	stdoutWriter := mergeWriters(cmd.Stdout, tf.stdout)
+	stderrWriter := mergeWriters(tf.stderr, &errBuf)
+
+	cmd.Stderr = nil
+	cmd.Stdout = nil
+
+	stdoutPipe, err := cmd.StdoutPipe()
+	if err != nil {
+		return err
+	}
+
+	stderrPipe, err := cmd.StderrPipe()
+	if err != nil {
+		return err
+	}
+
+	err = cmd.Start()
+	if ctx.Err() != nil {
+		return cmdErr{
+			err:    err,
+			ctxErr: ctx.Err(),
+		}
+	}
+	if err != nil {
+		return err
+	}
+
+	var errStdout, errStderr error
+	var wg sync.WaitGroup
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+		errStdoutCtx := ctx
+		// When cancellation delay is enabled, avoid piping cancellation to prevent
+		// cancellation from being interrupted due to broken pipe.
+		if cmd.WaitDelay > 0 {
+			errStdoutCtx = context.WithoutCancel(ctx)
+		}
+		errStdout = writeOutput(errStdoutCtx, stdoutPipe, stdoutWriter)
```

Comment:
question: Where do we wait for the subprocess exit? here (wait for pipe closed)?

Reply:
We have `wg.Wait()` at line 377 to wait for the buffer writing to finish. Once the graceful cancellation pass the timeout period, The command be sent kill signal, so this is guaranteed to be finished.

Do this answer your question?